File: Linked List/Double_Linked_list.cpp
#include <iostream>

using namespace std;
class node{
    public:
        int value;
        node *prev;
        node *next;

        node(int value){
            this->value = value;
            this->prev = nullptr;
            this->next = nullptr;
        }
};

class DoubleLinkedList{
    private: 
        node *start;
        node *end_p;
    public:
        DoubleLinkedList(){
            start = nullptr;
            end_p = nullptr;
        }
        void apped(int value){
            node *newnode = new node(value);
            if(start == nullptr){
                start = newnode;
                end_p = newnode;
            }else{
                end_p->next = newnode;
                newnode->prev = end_p;
                end_p = newnode;
            }
        }
        void dele(){
            if(end_p == nullptr){
                cout<<"List is Empty."<<endl;
                return;
            }
            node *temp = start;
            if(start == end_p){
                start = nullptr;
                end_p = nullptr;
            }else{
                end_p = end_p->prev;
                end_p->next = nullptr;
            }
            delete temp;
        }
        void del_between(int ind){
            node *ptr = start;
            node *p = start->next;
            for(int i=0; i<ind-1; i++){
                ptr = ptr->next;
                p = p->next;
            }
            ptr->next = p->next;
            if (p->next != nullptr){
                p->next->prev = ptr;
            }else{
                end_p = ptr;
            }
            delete p;
        }
        void display(){
            node *temp = start;
            while(temp!= nullptr){
                cout<<temp->value<<", ";
                temp = temp->next;
            }
            cout<<endl;
        }
        void display_reverse(){
            node *temp = end_p;
            while(temp != nullptr){
                cout<<temp->value<<", ";
                temp = temp->prev;
            }
            cout<<endl;
        }
};
int main(){
    DoubleLinkedList dl;
    dl.apped(55);
    dl.apped(4);
    dl.apped(10);
    dl.apped(78);
    dl.apped(103);

    cout<<"your linkelist: ";
    dl.display();
    dl.del_between(2);
    cout<<"after delete the index 2: ";
    dl.display();
    cout<<"Reverse the Linked list: ";
    dl.display_reverse();

}

File: Python/Linked_List/Double_Linked_list.py
class Node:
    def __init__(self, data):
        self.data = data
        self.prev = None
        self.next = None

class Double_Linked:
    def __init__(self):
        self.Head = None
    
    def append(self, data):
        new_node = Node(data)
        if self.Head is None:
            self.Head = new_node
            return
        last = self.Head
        while last.next:
            last = last.next
        last.next = new_node
        new_node.prev = last
    
    def prepend(self, data):
        new_node = Node(data)
        if not self.Head:
            self.Head = new_node
            return
        new_node.next = self.Head
        self.Head.prev = new_node
        self.Head = new_node
    
    def delete(self):
        if self.Head is None:
            print("Your Linked list is empty")
            return
        temp = self.Head
        if temp.next is None:
            self.Head = None
            return
        while temp.next is not None:
            temp = temp.next
        temp.prev.next = None
        
    
    def display(self):
        if self.Head is None:
            print("Your Linked list is empty")
            return
        temp = self.Head
        while temp is not None:
            print(temp.data, end=" -> ")
            temp = temp.next
        print("None")
    def print_backward(self):
        current = self.Head
        if not current:
            return
        
        while current.next:
            current = current.next
        
        while current:
            print(current.data, end=' -> ')
            current = current.prev
        print("None")

    

if __name__=="__main__":
    l1 = Double_Linked()
    l1.append(48)
    l1.append(89)

    l1.prepend(95)
    l1.append(67)
    l1.append(15)
    l1.display()
    print("after delete the node")
    l1.delete()
    l1.display()

    l1.print_backward()

        
        